lib/rubocop/cop/lint/non_deterministic_require_order.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # Dir[...] and Dir.glob(...) do not make any guarantees about
+      # the order in which files are returned. The final order is
+      # determined by the operating system and file system.
+      # This means that using them in cases where the order matters,
+      # such as requiring files, can lead to intermittent failures
+      # that are hard to debug. To ensure this doesn't happen,
+      # always sort the list.
+      #
+      # @example
+      #
+      #   # bad
+      #   Dir["./lib/**/*.rb"].each do |file|
+      #     require file
+      #   end
+      #
+      #   # good
+      #   Dir["./lib/**/*.rb"].sort.each do |file|
+      #     require file
+      #   end
+      #
+      # @example
+      #
+      #   # bad
+      #   Dir.glob(Rails.root.join(__dir__, 'test', '*.rb')).each do |file|
+      #     require file
+      #   end
+      #
+      #   # good
+      #   Dir.glob(Rails.root.join(__dir__, 'test', '*.rb')).sort.each do |file|
+      #     require file
+      #   end
+      #
+      class NonDeterministicRequireOrder < Cop
+        MSG = 'Sort files before requiring them.'
+
+        def_node_matcher :unsorted_dir_loop?, <<~PATTERN
+          (send (send (const nil? :Dir) {:[] :glob} ...) :each)
+        PATTERN
+
+        def_node_matcher :loop_variable, <<~PATTERN
+          (args (arg $_))
+        PATTERN
+
+        def on_block(node)
+          return unless node.children && node.children.length > 2
+
+          block_start = node.children[0]
+
+          return unless unsorted_dir_loop?(block_start)
+
+          var_name = loop_variable(node.children[1])
+
+          return unless var_name && var_is_required?(node, var_name)
+
+          add_offense(block_start)
+        end
+
+        def var_is_required?(node, var_name)
+          matcher = NodePattern.new("(send nil? :require (lvar :#{var_name}))")
+          node.each_descendant.find { |n| matcher.match(n) }
+        end
```

Comment:
There is no need to compile a new node pattern for each invocation. You can write this method like this:
```ruby
        def_node_search :var_is_required?, <<~PATTERN
          (send nil? :require (lvar %1))
        PATTERN
```
Node patterns can use arguments (`%1`) and `def_node_search` does the search in all descendants.

Reply:
Fixed, thanks! I figured there would be an easier way.